// @cosmjs/tendermint-rpc/build/jsonchecks.js
"use strict";
// This file contains runtime check for return values
// of a JSON API. They have a few things in common:
// 1. JSON inputs are expected. I.e. non-JSON values like Infinity or Uint8Array are not handled.
// 2. They have in input type we THINK the API returns but we don't trust that type.
// 3. The input value is returned unchanged if the check passes and an exception is thrown otherwise.
Object.defineProperty(exports, "__esModule", { value: true });
exports.jCheckSet = jCheckSet;
exports.jCheckBoolean = jCheckBoolean;
exports.jCheckString = jCheckString;
exports.jCheckNumber = jCheckNumber;
exports.jCheckArray = jCheckArray;
exports.jCheckObject = jCheckObject;
exports.jCheckNonEmptyString = jCheckNonEmptyString;
exports.jCheckNonZeroNumber = jCheckNonZeroNumber;
/**
 * A runtime checker that ensures a given value is set (i.e. not undefined or null)
 *
 * This is used when you want to verify that data at runtime matches the expected type.
 */
function jCheckSet(value) {
    if (value === undefined) {
        throw new Error("Value must not be undefined");
    }
    if (value === null) {
        throw new Error("Value must not be null");
    }
    return value;
}
/**
 * A runtime checker that ensures a given value is a boolean
 *
 * This is used when you want to verify that data at runtime matches the expected type.
 * This implies jCheckSet.
 */
function jCheckBoolean(value) {
    jCheckSet(value);
    if (typeof value !== "boolean") {
        throw new Error("Value must be a boolean");
    }
    return value;
}
/**
 * A runtime checker that ensures a given value is a string.
 *
 * This is used when you want to verify that data at runtime matches the expected type.
 * This implies jCheckSet.
 */
function jCheckString(value) {
    jCheckSet(value);
    if (typeof value !== "string") {
        throw new Error("Value must be a string");
    }
    return value;
}
/**
 * A runtime checker that ensures a given value is a number
 *
 * This is used when you want to verify that data at runtime matches the expected type.
 * This implies jCheckSet.
 */
function jCheckNumber(value) {
    jCheckSet(value);
    if (typeof value !== "number") {
        throw new Error("Value must be a number");
    }
    return value;
}
/**
 * A runtime checker that ensures a given value is an array
 *
 * This is used when you want to verify that data at runtime matches the expected type.
 * This implies jCheckSet.
 */
function jCheckArray(value) {
    jCheckSet(value);
    if (!Array.isArray(value)) {
        throw new Error("Value must be an array");
    }
    return value;
}
/**
 * A runtime checker that ensures a given value is an object in the sense of JSON
 * (an unordered collection of key–value pairs where the keys are strings)
 *
 * This is used when you want to verify that data at runtime matches the expected type.
 * This implies jCheckSet.
 */
function jCheckObject(value) {
    jCheckSet(value);
    if (typeof value !== "object") {
        throw new Error("Value must be an object");
    }
    // Exclude special kind of objects like Array, Date or Uint8Array
    // Object.prototype.toString() returns a specified value:
    // http://www.ecma-international.org/ecma-262/7.0/index.html#sec-object.prototype.tostring
    if (Object.prototype.toString.call(value) !== "[object Object]") {
        throw new Error("Value must be a simple object");
    }
    return value;
}
/**
 * Throws an error if value is not a string or is empty.
 *
 * Otherwise returns the value.
 *
 * This implies jCheckString.
 */
function jCheckNonEmptyString(value) {
    jCheckString(value);
    if (value.length === 0)
        throw new Error("Value must not be empty");
    return value;
}
/**
 * Throws an error if value is not a string or is empty.
 *
 * Otherwise returns the value.
 *
 * This implies jCheckNumber.
 */
function jCheckNonZeroNumber(value) {
    jCheckNumber(value);
    if (value === 0)
        throw new Error("Value must not be zero");
    return value;
}
