// cosmjs-types/cosmos/bank/module/v1/module.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Module = exports.protobufPackage = void 0;
/* eslint-disable */
const binary_1 = require("../../../../binary");
const helpers_1 = require("../../../../helpers");
exports.protobufPackage = "cosmos.bank.module.v1";
function createBaseModule() {
    return {
        blockedModuleAccountsOverride: [],
        authority: "",
        restrictionsOrder: [],
    };
}
exports.Module = {
    typeUrl: "/cosmos.bank.module.v1.Module",
    encode(message, writer = binary_1.BinaryWriter.create()) {
        for (const v of message.blockedModuleAccountsOverride) {
            writer.uint32(10).string(v);
        }
        if (message.authority !== "") {
            writer.uint32(18).string(message.authority);
        }
        for (const v of message.restrictionsOrder) {
            writer.uint32(26).string(v);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof binary_1.BinaryReader ? input : new binary_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseModule();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.blockedModuleAccountsOverride.push(reader.string());
                    break;
                case 2:
                    message.authority = reader.string();
                    break;
                case 3:
                    message.restrictionsOrder.push(reader.string());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromJSON(object) {
        const obj = createBaseModule();
        if (Array.isArray(object?.blockedModuleAccountsOverride))
            obj.blockedModuleAccountsOverride = object.blockedModuleAccountsOverride.map((e) => String(e));
        if ((0, helpers_1.isSet)(object.authority))
            obj.authority = String(object.authority);
        if (Array.isArray(object?.restrictionsOrder))
            obj.restrictionsOrder = object.restrictionsOrder.map((e) => String(e));
        return obj;
    },
    toJSON(message) {
        const obj = {};
        if (message.blockedModuleAccountsOverride) {
            obj.blockedModuleAccountsOverride = message.blockedModuleAccountsOverride.map((e) => e);
        }
        else {
            obj.blockedModuleAccountsOverride = [];
        }
        message.authority !== undefined && (obj.authority = message.authority);
        if (message.restrictionsOrder) {
            obj.restrictionsOrder = message.restrictionsOrder.map((e) => e);
        }
        else {
            obj.restrictionsOrder = [];
        }
        return obj;
    },
    fromPartial(object) {
        const message = createBaseModule();
        message.blockedModuleAccountsOverride = object.blockedModuleAccountsOverride?.map((e) => e) || [];
        message.authority = object.authority ?? "";
        message.restrictionsOrder = object.restrictionsOrder?.map((e) => e) || [];
        return message;
    },
};
